server/src/main/java/org/elasticsearch/action/admin/indices/template/post/SimulateIndexTemplateResponse.java

```diff
@@ -0,0 +1,124 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.action.admin.indices.template.post;
+
+import org.elasticsearch.action.ActionResponse;
+import org.elasticsearch.cluster.metadata.Template;
+import org.elasticsearch.common.Nullable;
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.ToXContentObject;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * Contains the information on what V2 templates would match a given index.
+ */
+public class SimulateIndexTemplateResponse extends ActionResponse implements ToXContentObject {
+
+    private static final ParseField TEMPLATE = new ParseField("template");
+    private static final ParseField OVERLAPPING_V1_TEMPLATES = new ParseField("overlapping_v1_templates");
```

Comment:
I'm a little on the fence about this, on one hand it's good we tell users what overlap, on the other hand, maybe we don't want to keep this around for a long time with the phrase "v1" in it.

What about including other V2 templates that could overlap here? And just calling it `overlapping_templates`? Or even expanding it to more information (that doesn't have to include the "vN" in the key name:

```
"overlapping": [
  { 
    "name": "template_1",
    "patterns": [...],
    "priority": 4,
    "type": "v1"
  }
]
```

What do you think? I'm wary of introducing the phrase "v1" for something we'll have to support for a very long time (past when "v1" templates are removed)

Reply:
I agree retiring the V1 concept will posses some problems. I do however question the utility of overlapping v2 templates though? (as, V2 templates don't overlap with V2 templates in the sense of some influencing the resulting mappings/settings of others, we have a clear winner)
I think we might be offloading a bit of work to the user to make some sense of that field and filter out the V2 templates.

Would it make sense to only call it "overlapping" and keep the current returned format and also keep only returning the V1 overlapping templates? Documenting this field accordingly.

When V1 templates are retired this field will always be empty so the clients will also not break.

Reply:
> Would it make sense to only call it "overlapping" and keep the current returned format and also keep only returning the V1 overlapping templates? Documenting this field accordingly.

Hmm.. what about calling it "overlapping" and including both V1 and V2 templates in the list? Eventually no more V1 templates will be in the list, but it could still be useful to see V2 templates that would have matched the index.

Reply:
Implemented this Lee, I think this is ready for review